### ISSUE TITLE
Wheel rotation speed for tracked vehicles is variable, depending on CPU load

### ISSUE THREAD

**Issue by handsm** [DarklightGames/DarkestHour#691]
Anyone who has set up the WheelRotationScale property for tracked vehicles will probably thought "I'm sure I did this before - how come it's visibly wrong?!". And has ended up setting the same vehicle more than once, or sometimes several people end up setting and re-setting it.

If you've done this you'll also probably have noticed that you can't put the game in slow motion(i.e. "slomo 0" in the console) to accurately match the wheel rotation speed to the tread movement. If you do the wheels continue to spin full speed, while the treads correctly move very slowly. That made me realise there is a fundamental problem in the RO wheel animation code - it takes no account of time elapsed between updates. CPU tick the wheels get rotated by a certain amount, proportionate to the current tread movement speed. But the time between ticks is highly variable, depending on how loaded the CPU is. So wheel rotation ends up variable, so at the same vehicle speed the wheels move slower when the CPU is heavily loaded and faster when it is ticking rapidly.

**Comment by handsm** [comment]
Tread and wheel animation happens on on network clients in multi-player. The server ignores all this stuff as irrelevant, while clients do an update every tick. Because the time between ticks is variable, when the Tick() native event is called it is passed the time elapsed since the last tick, which is a fraction of a second known as DeltaTime. Any functionality that is updating something like tread and wheel animation has to be built around DeltaTime. The problem here is the wheel animation ignores DeltaTime.

Tread movement is ok because treads are 'animated' by using a texture panner object and all you have to do is adjust its PanRate property to suit the vehicle's speed. The Unreal engine natively handles the movement each frame, factoring in elapsed time and the game speed (so in slow motion the treads move slower). But RO's wheel animation code just rotates the wheels by a calculated amount, based on vehicle speed but completely ignoring how much time has passed since the last update.

The fix is simple, but unfortunately all the WheelRotationScale settings for all tracked vehicles will become meaningless and have to be re-set. A nuisance but won't take more than an hour or two. And tbh the current settings are somewhat meaningless anyway, as they are only correct for a given tick rate.

Note this still leaves another underlying problem as individual wheel rotation speed is relative to that wheel's diameter, but the functionality applies a single rotation speed to all wheels on each side of the vehicle. Many vehicles have widely varying wheel sizes, e.g. drive sprocket and idler wheels may be larger than the road wheels, so they should rotate slower, while any return rollers are very small in comparison and should spin visibly faster. This could be done but imo it isn't very noticeable and the complexity it would add and the considerable time taken to set it up isn't justified by the issue.

**Comment by handsm** [comment]
Fixed in commit https://github.com/DarklightGames/DarkestHour/commit/a8bdc199816d866fa53bcb2ce03265dc650d2308.

The amount the wheels are rotated by is now multiplied by DeltaTime in Tick(). This means the values of each vehicle's WheelRotationScale now need to be far higher.

Changed WheelRotationScale to be a float instead of an int. Although it is used to set the wheel rotator pitch, which is an int, it goes into a rotation calculation with two other floats, so it makes sense to be a float. Also changed the LeftWheelsRotation and RightWheelsRotation variables from rotators to floats (and renamed from LeftWheelRot), since only the pitch component of the rotator changes and so that's all we need to record. Made it a float instead of the usual int for a rotator component so it retains floating point precision from tick to tick. The rotator that is applied is now just a local variable.

For now, I've just multiplied the previous WheelRotationScale by 65, which is perhaps a fairly average tick rate and seems to give a good result on a selection of vehicles. Later I'll work through them all and fine tune the settings. The main thing is that the wheel rotation speed setting is now reliably proportional to the vehicle and tread speed.
